docs: JS-PDK interface description (#16)
extism/proposals: EIP-009-js-pdk-interface-definition.md

> # JS-PDK Interface Defintion
> 
> ## Purpose
> 
> We need to support host functions in the JS PDK but the current way we define
> the interface, while working for exports, will not work for imports. Here is a rough
> outline of how the current system works:
> 
> First we compile quickjs to wasm, we call this the "engine" or "core" module.
> This module is a rust wasm app with quickjs embedded and has one custom function as an export `__invoke`.
> `__invoke` takes an index (n), it looks at all the javascript exports in your main module, sorts them alphabetically,
> then invokes the (nth) one.
> 
> The next step is to wizen and mutate this core module. The user passes in some javascript
> which we eval and wizen. We need need to expose the exports in the export section. We do this
> by evaling the js (in this compile step), reading the js module's exports, sorting them alphabetically,
> and adding them to the exports table in order. The code for these functions (we call thunks) are also generated 
> and just call `__invoke` with the export function id (where is sits in the order).
> 
> There are some downsides to doing this for exports, it's complicated and slow, but it also works fine.
> This will not work for imports because we'd need to come up with a very wacky way to programatically extract
> and type the imports from the user's js code. We can no longer magically assume we will know the wasm signature.
> Exports are easy because all Extism exports have the same signature. Imports do not. Furthermore this process
> of mutating a wasm module and adding some exports and functions works because it is a fairly additive process.
> Importing on the other hand requires re-aligning a lot of other tables. And if i were to continue custom coding
> that I'd be building a linker.

Reviewer: (These are just line edits as I followed the flow; contains some typo fixes.)

```suggestion
# JS-PDK Interface Definition

## Purpose

We need to support host functions in the JS PDK but the current way we define
the interface, while working for exports, will not work for imports. Here is a rough
outline of how the current system works:

1. We compile QuickJS to wasm to create the `engine` or `core` module. This
   module is a Rust Wasm app with QuickJS embedded; it has one custom export,
   `__invoke`. `__invoke` takes an index (`N`), looks at all the JS exports
   in your main module sorted alphabetically, then invokes the `N`th one.
2. Next, we use [Wizer] to partially evaluate the core module. The user passes
   in some JS which we evaluate as part of this process. We determine the Wasm
   exports by evaluating the JS in this partial evaluation step, reading the
   exports from the JS module, sorting them alphabetically, then adding them to
   our (Wasm) exports table. We generate code for these functions (or "thunks")
   which calls `__invoke` with the appropriate export function ID.

While it works fine for exports, there are some downsides to this approach.
It's complicated and slow, and does not work for imports. For imports, we'd
need to come up with a way to programmatically extract the imports and their
types from the user's JS code. We can't assume we'll know the Wasm signature.
Exports are easy because all Extism exports have the same signature, but
imports may have any valid Wasm signature. This technique works well for
exports in part because it is purely additive: we add new functions at the end
of the module; on the other hand, imports require re-aligning a lot of index
spaces. This path leads to writing a linker, which is out of scope for us right
now.
```

## Solution

What we need to start doing is have the programmer define for us their interface explicitly. They'll be able to
define the Exports and Imports in an IDL. We also need to use tooling to link the 
core module and the user code module so we don't fall down the trap of havign to build a linker.

Here is how I propose the new path will work. This may not be permanent but I think 
it will work for 1.0:

![js pdk pipeline](content/009-js-pdk-pipeline.png)

The programmer will define their interface in a typescript file `interface.d.ts`:

> ```typescript

Reviewer: Re: the imports and exports: it might be worth gesturing at future work here – noting that we might do basic mapping of higher-level types like `string` / `Record` and that we're leaving it at F32, I32, & friends for now.

> ```typescript
> // all extism exports have type () -> i32
> interface Exports {
>   myExport(): I32;
> }
> 
> // this interface is optional and types are  defined by you
> // to match your host functions
> interface Imports {
>   myHostFunc1(p: I64, q: I32): I64;
>   myHostFunc2(p: I64): I64;
> }
> ```

Reviewer: ```suggestion
```typescript

// modules named like WIT-style packages ("extism:host") declare host function imports
declare module 'extism:host' { // namespace
  interface user {
    export myHostFunction1(p: I32, q: I32): I32;
    export myHostFunction2(p: I32): I64;
  } // which tees us up to implement "import { user } from 'extism:host'"
}

// The "main" module declares exports
declare module 'main' {
  // all extism exports have type () -> i32
  export function greet(): I32;
}
```
```

and dump out to the final wasm after a few passes from wasm-opt.

### Rust to JS Bridge

Suppose we have 2 host functions and one export. The shim module that is generated
should look something like this (TODO this is still not accurate yet):

export-shim.wasm:

```wat
(module
  ; included by us for invoking the js runtime
  (import "coremod" "__invoke" (func (;0;) (type 0)))

  ; the generated thunk for the export
  (func (;1;) (type 0) (param i32) (result i32)
    local.get 0
    call 0)
  (export "myExport" (func 1)))
```

For exports, we can leave the behavior the same. For each export we will generate
a thunk function that calls `__invoke`.

For imports we have a bit of a challenge. The JS code, which is executed in the core
module, needs to be able to invoke these host functions.

The JS code (and underlying rust and c code) doesn't know about these functions, their names or their function
indexes at compile time. That's why the import shim will expose a call-indirect to the host functions.

TODO: this means all types need to be the same, perhaps we need to also expose the type parameter?

import-shim.wasm:

```wat
(module
  (type $int2int (func (param i32) (result i32)))

>   (import "coremode" "myHostFunc1" (func $myHostFunc1 (type $int2int)))
>   (import "coremode" "myHostFunc2" (func $myHostFunc2 (type $int2int)))

Reviewer: ```suggestion
  (import "coremod" "myHostFunc1" (func $myHostFunc1 (type $int2int)))
  (import "coremod" "myHostFunc2" (func $myHostFunc2 (type $int2int)))
```

  (table 2 funcref)
  (elem (i32.const 0) $myHostFunc1 $myHostFunc2)

  (func $callHostFunc (type $int2int)
    local.get 0
    (call_indirect (type $int2int) (i32.const 0))
  )

>   (export "__invokeHostFunc" (func $callHostFunc))

Reviewer: (We talked offline a bit about this, but noting it here for posterity) dispatching based on the different host import type signatures might be a little tricky (but not insurmountable!)

)
```

It exposes this indirect call with __invokeHostFunc which can be used by the
core quickjs engine module and invoked from javascript.

> 

Reviewer: (Another note from offline conversation, saved here for posterity!) There'll be an interesting middle step as we have to partially evaluate the module using Wizer while disabling the host function calls.

(This is probably retreading ground you've already walked, but it'd be interesting to consider whether or not users would want to "stub in" host fn calls made during this partial evaluation – as you noted, Wizer is able to implement stub functions.)
